staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -56,16 +57,28 @@ func WithTimeoutForNonLongRunningRequests(handler http.Handler, longRunning apir
 		ctx, cancel := context.WithCancel(ctx)
 		req = req.WithContext(ctx)
 
-		postTimeoutFn := func() {
+		postTimeoutFn := func(err error) {
 			cancel()
-			metrics.Record(req, requestInfo, metrics.APIServerComponent, "", http.StatusGatewayTimeout, 0, 0)
+
+			if err == nil {
+				// No error occur, timeout handler write 504 code and body successfully
+				metrics.Record(req, requestInfo, metrics.APIServerComponent, "", http.StatusGatewayTimeout, 0, 0)
+				return
+			}
+
+			// Otherwise, inner handler has already tried to write headers and body but timeout to finish to write all body,
+			// or timeout handler failed to write 504 response.
+			// So, actually it's not a completed 504 response to the client, the client received EOF.
+			// We need to distinguish this with logic processing timeout
+			// TODO(answer1991): if -1 code is ok here
+			metrics.Record(req, requestInfo, metrics.APIServerComponent, "", -1, 0, 0)
```

Comment:
I agree that none of the official HTTP response codes are appropriate here. In fact even nginx's 499, client closed the connection does not fit. However I do think we should probably try to use something more like a response code, rather that -1. This seems like a 5xx series error. What about something like a 594?

Reply:
I do not think this code is a response code, actually server had already write a response code to the client which maybe is 200 or any other official HTTP response code. So I set -1 here is just for metrics usage. Setting an unassigned HTTP response code(like 594) will cause conflict error if the code become standard official code in the feature.